src/main.cpp

```diff
@@ -4243,6 +4251,17 @@ bool static LoadBlockIndexDB()
 
             // Fall back to hardcoded Sprout value pool balance
             FallbackSproutValuePoolBalance(pindex, chainparams);
+
+            // If developer option -developersetpoolsizezero has been enabled on testnet or in regtest mode,
+            // override and set the in-memory size of shielded pools to zero.  An unshielding transaction
+            // can then be used to trigger and test the handling of turnstile violations.
+            if ((chainparams.NetworkIDString() == "test" || chainparams.NetworkIDString() == "regtest") &&
+                fExperimentalMode &&
+                mapArgs.count("-developersetpoolsizezero"))
+            {
+                pindex->nChainSproutValue = 0;
+                pindex->nChainSaplingValue = 0;
+            }
```

Comment:
This alters every single block, not just the chain tip, which might cause unexpected behaviour if a chain reorg were to occur while this flag was set (specifically, the reorged-to chain would be marked invalid non-deterministically). Non-blocking because I think this is sufficiently unlikely to occur, and if this happened on testnet, the developer could use the `reconsiderblock` RPC method to restore the main chain.